### ISSUE TITLE
Not that fast compared to warp-transducer on my machine

### ISSUE THREAD

**Issue by gmyofustc** [1ytic/warp-rnnt#2]
Hi 1ytic,
     thanks for share your great work. I test your code with that of warp-transducer, but the speed is not that fast, anything I get wrong? Looking forward to your response.

======
Speed Results
root@n130-018-083: python benchmark.py 
1: 1.23
1: 1.75
16: 2.43
16: 5.83
32: 33.22
32: 9.02
64: 58.27
64: 30.11
128: 141.53
128: 72.33
=====
Runing Config

run_benchmark_safe(loss, E=50, N=n, T=150, U=20, V=2000)

**Comment by 1ytic** [comment]
Thanks! What hardware do you have?

**Comment by gmyofustc** [comment]
@1ytic hi, thanks for you quick response. I use Tesla V100 GPU with CUDA10.1 and Pytorch version of 1.3

**Comment by 1ytic** [comment]
Here is my test on V100:

`run_benchmark(loss, E=100, N=n, T=150, U=40, V=28)`
1: 0.12
1: 0.75
16: 0.44
16: 1.47
32: 0.53
32: 2.27
64: 0.76
64: 3.34
128: 0.89
128: 5.48

`run_benchmark_safe(loss, E=10, N=n, T=150, U=20, V=5000)`
1: 0.70
1: 1.79
16: 2.52
16: 9.27
32: 4.37
32: 17.22
64: 8.08
64: 33.38

`run_benchmark_safe(loss, E=10, N=n, T=1500, U=300, V=50)`
1: 0.92
1: 10.12
16: 3.71
16: 44.50
32: 6.63
32: 86.4

This is consistent with 1080Ti test.

Maybe you built warp_rnnt in a debug mode. How did you install it? Did you modify the benchmark.py script?

**Comment by gmyofustc** [comment]
Hi, I install warp-rnnt use your default settings. With your suggestion, I run more test and find that:
warp-rnnt has great speed up when the computation cost is small(say small batch_size, or small T,U,V， but the speed up will decrease seriously when the computation comes up)

This is my configuration:
`run_bentch_mark_safe(loss, E=10, N=n, T=400, U=50, V=6000)`
the speed results are:
1: 1.30
1: 10.69
2: 1.63
2: 14.63
4: 2.57
4: 27.26
8: 60.62
8: 50.15
16: 165.08
16: 209.90
32: 234.24
32: OOM
+++++++++++++++++++++++
when batch_size increase from 4->8, warp-rnnt will cost much more time, I test this lots of times, and the results are consistent

**Comment by 1ytic** [comment]
Indeed, with long utterances and a huge vocabulary calculations slow down. Unfortunately, I cannot reproduce your degradation problem between 4 and 8 batch_size.

On the other hand, with a huge vocabulary, I highly recommend using [gather](https://github.com/1ytic/warp-rnnt/blob/c8da7e029750d10252e98c873ba333cd639918f6/pytorch_binding/warp_rnnt/__init__.py#L33) flag. Here is my test with gather flag on/off:

`run_benchmark_safe(loss, E=5, N=n, T=400, U=50, V=6000)`
1: 1.13 <- gather=True
1: 1.65 <- gather=False
1: 6.01 <- warprnnt_pytorch
2: 1.48 <- gather=True
2: 2.62
2: 9.63
4: 2.60 <- gather=True
4: 4.26
4: 17.61
8: 4.72 <- gather=True
8: 7.94
8: 33.38
16: 8.25 <- gather=True
16: OOM
16: 64.96

Please let me know if this solves your issue. Good luck.

**Comment by gmyofustc** [comment]
Thanks for your suggestion, I used the gather flag, this indeed reduced the memory cost and gives same speedup, now I can training my network with your repo and all things go well, thanks again for your great job.
